packages/typescript-koa-runtime/src/server.ts

```diff
@@ -1,51 +1,137 @@
 export type StatusCode3xx = IntRange<300, 399> // `3${number}${number}`
 export type StatusCode4xx = IntRange<400, 499> // `4${number}${number}`
 export type StatusCode5xx = IntRange<500, 599> // `5${number}${number}`
-export type StatusCode = StatusCode1xx | StatusCode2xx | StatusCode3xx | StatusCode4xx | StatusCode5xx
+export type StatusCode =
+  | StatusCode1xx
+  | StatusCode2xx
+  | StatusCode3xx
+  | StatusCode4xx
+  | StatusCode5xx
 
-export type Response<Status extends StatusCode, Type> = { status: Status, body: Type }
+export type Response<Status extends StatusCode, Type> = {
+  status: Status
+  body: Type
+}
 
 export type ServerConfig = {
+  /** set to "disabled" to disable cors middleware, omit or pass undefined for defaults */
+  cors?: "disabled" | Cors.Options | undefined
+
+  /**
+   * set to "disabled" to disable body parsing middleware, omit or pass undefined for defaults.
+   *
+   * if disabling, ensure you pass a body parsing middlware that places the parsed
+   * body on `ctx.body` for request body processing to work.
+   **/
+  body?: "disabled" | KoaBodyMiddlewareOptions | undefined
+
+  /**
+   *
+   * useful for mounting logging, alternative body parsers, etc
+   */
   middleware?: Middleware[]
+
+  /**
+   * the router to use, normally obtained by calling the generated `createRouter`
+   * function
+   */
   router: Router
-  port: number
+
+  /**
+   * the port to listen on, a randomly allocated port will be used if none passed
+   * alternatively ListenOptions can be passed to control the network interface
+   * bound to.
+   */
+  port?: number | ListenOptions
 }
 
-export function startServer(config: ServerConfig): Server {
+/**
+ * Starts a Koa server and listens on `port` or a randomly allocated port if none provided.
+ * Enables CORS and body parsing by default. It's recommended to customize the CORS options
+ * for production usage.
+ *
+ * If you need more control over your Koa server you should avoid calling this function,
+ * and instead mount the router from your generated codes `createRouter` call directly
+ * onto a server you have constructed.
+ */
+export async function startServer({
+  middleware = [],
+  cors = undefined,
+  body = undefined,
+  port = 0,
+  router,
+}: ServerConfig): Promise<{
+  app: Koa
+  server: Server
+  address: AddressInfo
+}> {
   const app = new Koa()
 
-  app.use(cors())
-  app.use(koaBody())
+  if (cors !== "disabled") {
+    app.use(Cors(cors))
+  }
 
-  const middleware = config.middleware ?? []
-  middleware.forEach(it => app.use(it))
+  if (body !== "disabled") {
+    app.use(KoaBody(body))
+  }
 
-  app.use(config.router.allowedMethods())
-  app.use(config.router.routes())
+  middleware.forEach((it) => app.use(it))
 
-  const server = app.listen(config.port, () => {
-    const address = server.address()
-    console.info("server listening", {address})
-  })
+  app.use(router.allowedMethods())
+  app.use(router.routes())
+
+  return new Promise((resolve, reject) => {
```

Comment:
Pretty ugly use of `new Promise` and several `try`-`catch` blocks, but whilst I don't think `sever.address()` throws right now, there's a `FIXME` comment suggesting it [might in future](https://github.com/nodejs/node/blob/main/lib/net.js#L897), and doing this means that any exceptions will definitely stay in the promise chain correctly.

Also starts handling errors like `EADDRINUSE` properly, as these get emitted as an `error` event